### PR TITLE
Exclusion lists for synced folders

### PR DESCRIPTION
**Pull request by javiserrano**
This pull request is intended to add support to exclude relative paths from the synchronization instead of just file / folder names.
Additionally, it introduces a new exclusion method based on regular expressions. The new method can be enabled with `USE_PCRE`, but it's not expected to be 100% stable yet and it's very problable that the interface of the object `MegaRegExp` changes in few time to support different types of regular expressions and a whitelist.

### CODE REVIEW

File: src/megaapi_impl.cpp [hunk 1]
```diff
@@ -15993,6 +16209,230 @@ void MegaSyncPrivate::setState(int state)
...
+    return regExps.size();
+}
+
+const char * MegaRegExpPrivate::getRegExp(int index)
+{
+    return regExps.at(index).c_str();
+}
+
+/**
+ * @brief Checks if the given regular expression is
```
Comment (polmr) [inline src/megaapi_impl.cpp]: *is correct

File: include/megaapi.h [hunk 2]
```diff
@@ -7945,6 +8044,36 @@ class MegaApi
...
+        MegaSync *getSyncByNode(MegaNode *node);
+
+        /**
+         * @brief getSyncByPath Get the syncrhonization associated with a local path
+         * @param localPath Root local path of the synchronization
+         * @return Synchronization with the specified root local path
+         */
+        MegaSync *getSyncByPath(const char *localPath);
+
+#ifdef USE_PCRE
```
Comment (javiserrano) [inline include/megaapi.h]: Name changed 👍

Comment (javiserrano) [inline include/megaapi.h]: I agree, it's already planned to add support for wildcards in MegaRegExp.

File: src/megaapi_impl.cpp [hunk 3]
```diff
@@ -6401,6 +6437,14 @@ bool MegaApiImpl::isSyncable(const char *path, long long size)
 
     bool result = false;
     sdkMutex.lock();
+    if (size >= 0)
+    {
+        if (!is_syncable(size))
+        {
+            return false;
```
Comment (polmr) [inline src/megaapi_impl.cpp]: missing mutex unlock

**Pull request merged**